### DISCUSSION TITLE
Pre-processing of some options

### DISCUSSION THREAD

**Discussion by pawamoy** [DanCardin/cappa#162] [Show and tell]
I'm rewriting an argparsed-based CLI with Cappa. In the previous `main` function, I was first parsing arguments into an argparse namespace, then using some options of this namespace to configure logging.

```python
def main(args: list[str] | None = None) -> int:
    parser = get_parser()
    opts = parser.parse_args(args=args)
    configure_logging(opts.log_level, opts.log_path, allow="pypiserver")
    if not opts.subcommand:
        parser.print_usage(file=sys.stderr)
        return 1
    return commands[opts.subcommand](opts)
```

After a bit of twiddling, here's how I achieved the same thing with Cappa:

```python
# The following log options are added to every command.
@dataclass(kw_only=True)
class LogLevelPathOptions:
    """Log level and path options."""

    log_level: An[
        Literal["TRACE", "DEBUG", "INFO", "SUCCESS", "WARNING", "ERROR", "CRITICAL"],
        cappa.Arg(short="-L", long=True, parse=str.upper),
        Doc("Log level to use when logging messages."),
    ] = "INFO"

    log_path: An[
        str | None,
        cappa.Arg(short="-P", long=True),
        Doc("Write log messages to this file path. (Default: stderr)"),
    ] = None

    # This method actually configures logging with the provided CLI arguments.
    def process(self) -> None:  # noqa: D102
        configure_logging(self.log_level, self.log_path, allow="pypiserver")


# The logic for calling these `process()` methods is implemented in a mixin.
class ProcessOptions:
    def process_options(self) -> None:  # noqa: D102
        for parent in self.__class__.__mro__:
            if parent is ProcessOptions:
                continue
            if hasattr(parent, "process"):
                parent.process(self)


# Now just for convenience I declare a "common options" class
# that holds all my common options, to avoid repetition later.
# Note how it inherits my "options pre-processing" mixin
# (as well as other classes defining some options).
@dataclass(kw_only=True)
class CommonOptions(HelpOption, ConfPathOption, LogLevelPathOptions, ProcessOptions):
    """Common options for all commands."""


# Finally all my commands can inherit from `CommonOptions`,
# and call the mixin's `process_options()` method in their `__call__` method.
@cappa.command(
    name="start",
    help="Start the server.",
    description="Start the server in the background.",
)
@dataclass(kw_only=True)
class CommandServerStart(CommonOptions, DistDirOption, PortOption):
    """Command to start the server."""

    def __call__(self) -> int:  # noqa: D102
        self.process_options()
        start_server(dist_dir=self.dist_dir, port=self.port)
        return 0
```

And my `main` function doesn't have to do any special processing itself:

```python
def main(args: list[str] | None = None) -> int:
    output = cappa.Output(error_format=f"[bold]{NAME}[/]: [bold red]error[/]: {{message}}")
    return cappa.invoke(CommandMain, argv=args, output=output, backend=cappa.backend, completion=False, help=False)
```

**Comment by DanCardin** [comment]
I tend to do this sort of stuff in invoke dependencies

```python
from __future__ import annotations

from dataclasses import dataclass
from typing import Annotated as An
from typing import Literal

import cappa


def process(command: Command):
    print(f"configure_logging: {command.log_level}")


@dataclass(kw_only=True)
class Command:
    log_level: An[Literal["TRACE", "DEBUG"], cappa.Arg(short="-L")] = "INFO"
    subcommands: cappa.Subcommands[Foo | Bar]


@dataclass
class Foo:
    def __call__(self):
        print("foo")


@dataclass
class Bar:
    def __call__(self):
        print("bar")


cappa.invoke(Command, deps=[process])
```

If there's a direct dependence on the output in downstream deps, you can just depend on them directly, or if (with logging) you'd want to do it on **all** subcommands regardless of context, and dont want to explicitly call out to it in each command's impl.

---
Not relevant to this example, but something that occurs to me.

Although a potential drawback here is if the dep is only utilized in a subset of commands (e.g. `git remote [set-url, list, whatever]`. one level removed from the top, you'd (currently) have to resort to calling out to the function yourself. it's probably an indicator that `@command(deps=[])` might be worth looking into

**Reply by pawamoy** [reply]
Thanks, that looks much better indeed, and fits my use-case, as *every* command has these options and should configure logging before anything :slightly_smiling_face: I'll try it out!

**Reply by pawamoy** [reply]
I tried doing this:

```python
    dependencies = [lambda command: configure_logging(command.log_level, command.log_path, allow="pypiserver")]
    return cappa.invoke(CommandMain, argv=args, output=output, deps=dependencies, completion=False, help=False)
```

...but it fails with:

```pytb
Traceback (most recent call last):
  File "/media/data/dev/pypi-insiders/.venv/lib/python3.12/site-packages/cappa/invoke.py", line 154, in resolve_callable
    global_deps = resolve_global_deps(deps, implicit_deps)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/media/data/dev/pypi-insiders/.venv/lib/python3.12/site-packages/cappa/invoke.py", line 182, in resolve_global_deps
    value = fulfill_deps(dep.callable, implicit_deps)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/media/data/dev/pypi-insiders/.venv/lib/python3.12/site-packages/cappa/invoke.py", line 324, in fulfill_deps
    raise InvokeResolutionError(
cappa.invoke.InvokeResolutionError: `command: <empty>` is not a valid dependency for Dep(<lambda>).

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/media/data/dev/pypi-insiders/.venv/bin/pypi-insiders", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "/media/data/dev/pypi-insiders/src/pypi_insiders/cli/__init__.py", line 671, in main
    return cappa.invoke(CommandMain, argv=args, output=output, deps=dependencies, completion=False, help=False)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/media/data/dev/pypi-insiders/.venv/lib/python3.12/site-packages/cappa/base.py", line 143, in invoke
    resolved, global_deps = resolve_callable(
                            ^^^^^^^^^^^^^^^^^
  File "/media/data/dev/pypi-insiders/.venv/lib/python3.12/site-packages/cappa/invoke.py", line 159, in resolve_callable
    raise InvokeResolutionError(
cappa.invoke.InvokeResolutionError: Failed to invoke <class 'pypi_insiders.cli.CommandServerStatus'> due to resolution failure.
```

...so I suppose lambdas, or more generally `<locals>`-scope dependencies aren't supported :slightly_smiling_face:

**Reply by pawamoy** [reply]
Oh, with a regular function it works, so only lambdas seem to be an issue. Want a proper issue?

**Reply by pawamoy** [reply]
Ah, it's because `command` is not annotated in the lambda.

**Reply by pawamoy** [reply]
Hmmm and another issue with the dependencies approach is that they're fired only for the type I annotate `command` with, in this case `CommandMain` :thinking: Maybe that's what you meant with:

> If there's a direct dependence on the output in downstream deps, you can just depend on them directly

...and I misunderstood.

**Reply by DanCardin** [reply]
you already arrived at it, but yea invoke deps all need annotations, lambdas should work so long as there's an annotation.

---

idk if your situation is atypical or not, but i would expect every CLI to stem from a central object that represents the root of the CLI. e.g. the cls that's being `invoke`'d against.

...and having written that I assume I understand your issue. You're manually propagating `--log-level` down into all the subcommands, rather than relying on a single centralized `--log-level` option on the root cli? (i'd wonder what you'd expect the behavior to be if someone supplies both a top-level and lower-level option?)

My suggestion was predicated on the assumption you'd have a central option and dispatching to `command: CommandMain` would correctly yield you `command.log_level` as the thing to read. You **could** traverse the command tree from the root with a predictable enough CLI shape, and perhaps that would yield you more robust support for someone supplying the argument at different levels of subcommand.

---

Something I've thought about before is annotating `Self` to mean "the current invoked command". And I wonder if this would address your problem, if you explicitly want it to only consider the lowest level selected subcommand's `--log-level`?

e.g. `lambda self: Self: configure_logging(self.log_level, ...`. I just tried a adding it to invoke (annoyingly needing to deal with `typing_extensions.Self` and `typing.Self` because they're not the same), and it seems to be fairly straightforward to add.

**Reply by pawamoy** [reply]
Good analysis. I indeed add these same options to every command and subcommand. I do this because I find it more ergonomic for the end-user. I find it annoying to have to move an option at the front of the line because it's a global option. I'm not sure what's the convention here :thinking: 

With the options only on the main command:

```console
% mycli command subcommand
does the thing, but I want more logs, hit arrow-up and...
% mycli command subcommand -L DEBUG
error: unknown option -L
```

> (i'd wonder what you'd expect the behavior to be if someone supplies both a top-level and lower-level option?)

I think I don't expect anything as I'd consider this a user error. Cappa won't fail and that's fine, and it will probably only configure logging with the last occurrence of `-L`, which is fine too.

**Reply by pawamoy** [reply]
If you take a look at [uv](https://pypi.org/project/uv/), it does support the same global options at any level:

```bash
uv -q python list
uv python -q list
uv python list -q
```

That's what I'm trying to do too :smile:

**Reply by DanCardin** [reply]
Right i guess i'm just saying that my reading of your original example wouldn't behave that way because it doesn't distinguish between options at different levels, unless `configure_logging` is doing something internally (but i can't see how it would, because the default is INFO, so it can't tell between the default and a lack of an input?

And rereading your original comparison with argparse, I would assume it's that argparse writes all the results to a flat namespace, so whichever one is parsed last will happen to win (i assume).

And so, if I were to implement usage of `Self` (which i dont see a downside of, either way), then i'd suggest **that**, but then it would **only** yield/read the selected subcommand's log_level, and e.g. `uv python list -q` would work, but `uv -q python list` wouldn't do anything.

Whereas my original suggestion with `command: CommandMain` would be the opposite, although technically you could then traverse the command heirarchy `command.subcommand.subcommand.log_level` (given a normal-enough command structure) to implement a heirarchy lookup yourself.

The third (or fourth, after your original impl) option might be something like what uv/clap are ultimately using: https://docs.rs/clap/latest/clap/struct.Arg.html#method.global.

**Reply by DanCardin** [reply]
...also seems like the type checkers dont like this use of Self because it wont end up being in the context of a class definition. So, that's mildly unfortunate.

**Reply by pawamoy** [reply]
Getting inspiration from Clap sounds like a good idea :relaxed: Also, yeah, I thought using `Self` would indeed not please type checkers. Honestly it would kinda feel like a typing hack to use `Self` for this use-case :slightly_smiling_face: 

In the end I'm using the initial solution I showed (`process` method on options, `process_options` method on commands, inherited thanks to a mixin class). I worried it would be annoying for reusability in other CLIs (I'm planning to use one project's command into another project, as a subcommand :rocket:), but the other CLIs can just subclass the option to empty the `process` method, or subclass the command to empty the `process_options` method, so that's fine.

If you ever find an elegant way to support this use-case in a more straight-forward fashion, let me know! Otherwise it's all good for me :slightly_smiling_face:

**Reply by DanCardin** [reply]
So fwiw https://github.com/DanCardin/cappa/tree/dc/invoke-self. I'll probably look into global options next, also. seems like that'd also solve your problem by defining it once at the top level with `Arg(global=True)` then it would always get written back to the root object regardless of where it's supplied.

**Comment by pawamoy** [comment]
I simplified things a bit:

```python

@dataclass(kw_only=True)
class CommonOptions(LogLevelPathOptions, ConfPathOption, HelpOption):
    """Common options for all commands."""


@dataclass(kw_only=True)
class BaseCommand(CommonOptions):
    """Base command class."""

    def run(self) -> int:
        """Run the command."""
        raise NotImplementedError

    def __call__(self) -> int:  # noqa: D102
        configure_logging(self.log_level, self.log_path, allow="pypiserver")
        return self.run()
```

Now each command inherits from `BaseCommand`, and so will configure logging when called.

Why the `run` method though? you might ask. Well, it's simply so I don't have to keep calling `super().__call__()` in each command's `__call__` method (that I now don't have to implement, instead I implement `run`).
